tests/rptest/tests/delete_records_test.py

```diff
@@ -552,6 +560,51 @@ def test_delete_records_compacted_topic(self, cloud_storage_enabled):
         ):
             self.rpk.trim_prefix(TEST_COMPACTED_TOPIC_NAME, 0, [0])
 
+    @cluster(num_nodes=3)
+    @parametrize(cloud_storage_enabled=True)
+    @parametrize(cloud_storage_enabled=False)
+    def test_delete_records_topic_policy_change(self, cloud_storage_enabled):
```

Comment:
Nice test, I see in a following commit you also made a another test to ensure retention didn't break as well. I also think it might make sense to add conditional checks within the eviction stm itself.

I took a look at where the events are driven from, looks like it comes from a method called `do_gc` in `disk_log_impl.cc`. There is a check for `is_collectible()` before raising an event however there is a case when cloud storage is enabled where a storage event will arise. I'm not sure how that would affect things if the code were to go down that path and is_collectible is false.

For this reason i think we should edit the log eviction stm to also check if is collectible is true before applying eviction events from storage. I think a good place to do this might be in: `log_eviction_stm::monitor_log_eviction`. I know @andrwng also had opinions about refactoring that eviction background loop maybe he can chime in here too.

Reply:
disk_log_impl looks like the source of truth from my digging. archival is driven by the same events. I believe the current event flow to be sane.

You made me think about at least adding an assert in `log_eviction_stm`. But, if I understand the interplay between STMs and `raft::consensus`, and thus the topic config, is that they live on different timelines (concurrently spawned futures). Simply adding a `is_collectible` check in the `log_eviction_stm` might lead to race conditions where the `raft::consensus` is ahead (i.e. with compact.policy=compact), but the stm itself is processing an eviction raised earlier when the `compact.policy=delete`.

Reply:
The eviction stm isn't synchronized against anything in its current state anyway. When the storage layer alerts the eviction stm to evict, it does its best to truncate up to that point in a background fiber. You can also delete at a random offset with the DeleteRecords API.

Reply:
>disk_log_impl looks like the source of truth

`disk_log_impl` is where storage-centric truncations start (e.g. to apply local storage retention rules, or to aggressively GC because we're low on space). As Rob mentions, the delete-records case OTOH begins in `log_eviction_stm` and is user-driven (that's the `partition::prefix_truncate()` call). Given both cases are handled, this looks like it should work.

>Simply adding a is_collectible check in the log_eviction_stm might lead to race conditions

I'm not following this. Could you clarify? If we check `is_collectible` in the monitor, we wouldn't schedule any evictions (by setting `_storage_eviction_offset`, which is ultimately what drives Raft to write its snapshot). 

As Rob mentioned, I was originally thinking we could toggle the truncation loop in `log_evction_stm`, rather than blocking the scheduling of evictions. I've had a change of heart here: it seems worthwhile to keep the eviction loop running, even if we have a compact-only topic (by default, it'd just wait on a condition variable, which is cheap). With the new space management code, it seems entirely reasonable for us to schedule local GC on a compact-only topic if we're low on space, provided we've already uploaded the data (guaranteed by the max collectible offset check in the log_eviction_stm). For the most part, I'm okay with this change as is, but I'd request we follow it up with this behavior + tests.

Reply:
I updated the code with an extra check inside log_eviction_stm as requested. I was confused about the race condition I mentioned due to some wrong assumptions about how eviction works. Specifically, I thought that disk eviction is communicated via raft too and could lead to diverging replicas if stm is behind raft (which seems to be possible in theory). Since the actions are node local, any possible race conditions here aren't relevant.

Regarding the behavior for compact only topics (not evicting tiered storage segments under storage pressure), there is now a test in place to guard against changing the behavior (added via https://github.com/redpanda-data/redpanda/pull/13935) and an open issue to discuss separately whether we should change it (https://github.com/redpanda-data/redpanda/issues/14015). I don't see why we would block this bugfix PR on another issue not related to the origin bug report.

@andrwng @graphcareful did I understand your requests correctly?

Reply:
lgtm